docs/src/guidelines.md

```diff
@@ -0,0 +1,97 @@
+# [Guidelines for creating models](@id Guidelines)
+
+These are guidelines for the creation of models using NLPModels to help keeping the models uniform, and for future reference in the creation of solvers.
+
+## Bare minimum
+
+Your model should derive from `AbstractNLPModel` or some other abstract class derived from it.
+It is mandatory that it has a `meta :: NLPModelMeta` field, storing all the relevant problem information.
+The minimum information that should be set is
+
+- `nvar`: Number of variables
+- `x0`: Starting point (defaults to `zeros`)
+- `lvar`, `uvar`: Bounds on the variables (default to `(-∞,∞)`)
+- `ncon`: Number of constraints (defaults to `0`)
+- `lcon`, `ucon`: Bounds on the constraints (default to `(-∞,∞)`)
+- `nnzh`: The length of the vectors used to store a triangle of the Hessian matrix in triplet format (defaults to `nvar * (nvar + 1) / 2`
+- `nnzj`: The length of the vectors used to store the Jacobian matrix in triplet format (default to `nvar * ncon`)
+
+The model also need to provide `Counters` information, the easiest way being defining `counters :: Counters`.
+For instance:
+```julia
+mutable struct MyModel <: AbstractNLPModel
+  meta :: NLPModelMeta
+  counters :: Counters
+end
+```
+
+If a model does not implement `counters`, then it needs to define `neval_xxx` functions, such as in the following macro:
+```julia
+for counter in fieldnames(Counters)
+  @eval begin
+    $counter(nlp :: MyModel) = SOMETHING
+  end
+end
```

Comment:
Just a thought here. We could easily define a macro `@default_counters` (or something similar) so that the line
```julia
@default_counters MyModel inner_model
```
expands to
```julia
for counter in fieldnames(Counters)
   @eval begin
     $counter(nlp :: MyModel) = $counter(nlp.inner_model)
   end
 end
```

Reply:
I'm not sure it's worth the trouble.

Reply:
I noticed that I also need some additional counter functions, such as `increment!` and `sum_counters`, so I've written the macro to check how beneficial it is. Let me know what you think.